tensornetwork/backends/jax/jitted_functions.py

```diff
@@ -55,83 +62,100 @@ def jax_lanczos(matvec, arguments, init, ncv, neig, landelta, reortho):
+          reortho, lambda x: iterative_classical_gram_schmidt(
+              previous_vector,
+              (i > jax.numpy.arange(ncv + 2))[:, None] * krylov_vectors),
+          lambda x: previous_vector, None)
+
+      beta = jax.numpy.linalg.norm(previous_vector)
+      normalized_vector = previous_vector / beta
+      Av = matvec(jax.lax.reshape(normalized_vector, shape), *arguments)
+      alpha = jax.numpy.vdot(normalized_vector, Av, precision=precision)
+      alphas = alphas.at[i - 1].set(alpha)
+      betas = betas.at[i].set(
+          beta)
+      next_vector = jax.lax.cond(
+          reortho, lambda x: Av, lambda x: jax.numpy.reshape(
+              jax.numpy.ravel(Av) - jax.numpy.ravel(normalized_vector) * alpha -
+              krylov_vectors[i - 1] * beta, Av.shape), None)
+      krylov_vectors = krylov_vectors.at[i, :].set(
+          jax.numpy.ravel(normalized_vector))
+      krylov_vectors = krylov_vectors.at[i + 1, :].set(
+          jax.numpy.ravel(next_vector))
+
+      return [krylov_vectors, alphas, betas, i + 1]
 
     def cond_fun(vals):
-      _, _, _, _, _, _, norm, threshold, iteration, maxiteration = vals
-
-      def check_thresh(check_vals):
-        val, thresh = check_vals
-        return jax.lax.cond(val < thresh, False, lambda x: x, True, lambda x: x)
-
-      return jax.lax.cond(iteration <= maxiteration, [norm, threshold],
-                          check_thresh, False, lambda x: x)
-    #TODO (mganahl): check if this runs on TPU (dtype issue)
-    numel = np.prod(init.shape).astype(np.int32)
-    krylov_vecs = jax.numpy.zeros((ncv + 1, numel), dtype=init.dtype)
-    norms = jax.numpy.zeros(ncv, dtype=init.dtype)
-    diag_elems = jax.numpy.zeros(ncv, dtype=init.dtype)
-
-    norms = jax.ops.index_update(norms, jax.ops.index[0], 1.0)
-
-    norms_dtype = jax.numpy.real(jax.numpy.empty((0, 0),
-                                                 dtype=init.dtype)).dtype
-    initvals = [
-        init, krylov_vecs, norms, diag_elems, matvec, arguments,
-        norms_dtype.type(1.0), landelta, 1, ncv
-    ]
-    output = jax.lax.while_loop(cond_fun, body_lanczos, initvals)
-    final_state, krylov_vecs, norms, diags, _, _, _, _, it, _ = output
-    krylov_vecs = jax.ops.index_update(krylov_vecs, jax.ops.index[it, :],
-                                       jax.numpy.ravel(final_state))
-
-    A_tridiag = jax.numpy.diag(diags) + jax.numpy.diag(
-        norms[1:], 1) + jax.numpy.diag(jax.numpy.conj(norms[1:]), -1)
+      betas, i = vals[-2], vals[-1]
+      norm = betas[i - 1]
+      return jax.lax.cond(i <= ncv, lambda x: x[0] > x[1], lambda x: False,
+                          [norm, landelta])
+
+    numel = np.prod(shape).astype(np.int32)
+    # note: ncv + 2 because the first vector is all zeros, and the
+    # last is the unnormalized residual.
+    krylov_vecs = jax.numpy.zeros((ncv + 2, numel), dtype=dtype)
+    # NOTE (mganahl): initial state is normalized inside the loop
```

Comment:
I personally don't like calling the vector a "state" in these functions, but again it's a matter of taste.